homeassistant/components/generic/config_flow.py

```diff
@@ -515,3 +537,97 @@ async def get(self, request: web.Request, flow_id: str) -> web.Response:
             CAMERA_IMAGE_TIMEOUT,
         )
         return web.Response(body=image.content, content_type=image.content_type)
+
+
+class PreviewStream:
+    """A wrapper around the stream object to automatically close unused streams."""
+
+    def __init__(self, stream: Stream) -> None:
+        """Initialize the object."""
+        self.stream = stream
+        self._deferred_stop = None
+
+    async def start(self, timeout=600):
+        """Start the stream with a timeout."""
+
+        async def _timeout() -> None:
+            _LOGGER.debug("Starting preview stream with timeout %ss", timeout)
+            await asyncio.sleep(timeout)
+            _LOGGER.info("Preview stream stopping due to timeout")
+            await self.stream.stop()
+
+        await self.stream.start()
+        self._deferred_stop = self.stream.hass.async_create_task(_timeout())
+
+    async def stop(self):
+        """Stop the stream."""
+        if not self._deferred_stop.done():
+            self._deferred_stop.cancel()
+        await self.stream.stop()
+
+    def add_provider(self, provider):
+        """Add a provider to the stream."""
+        return self.stream.add_provider(provider)
+
+    def endpoint_url(self, fmt: str) -> str:
+        """Return the endpoint URL."""
+        return self.stream.endpoint_url(fmt)
+
+
+@websocket_api.websocket_command(
+    {
+        vol.Required("type"): "generic_camera/start_preview",
+        vol.Required("flow_id"): str,
+        vol.Required("flow_type"): vol.Any("config_flow"),
+        vol.Required("user_input"): dict,
+    }
+)
+@websocket_api.async_response
+async def ws_start_preview(
+    hass: HomeAssistant,
+    connection: websocket_api.ActiveConnection,
+    msg: dict[str, Any],
+) -> None:
+    """Generate websocket handler for the camera still/stream preview."""
+    _LOGGER.debug("Generating websocket handler for generic camera preview")
+    ha_still_url = None
+    ha_stream_url = None
+
+    flow_id = msg["flow_id"]
+    flow = cast(
+        GenericIPCamConfigFlow,
+        hass.config_entries.flow._progress.get(flow_id),  # noqa: SLF001
+    ) or cast(
+        GenericOptionsFlowHandler,
+        hass.config_entries.options._progress.get(flow_id),  # noqa: SLF001
```

Comment:
is the options flow not used yet? If it is not used yet then remove for now and add in follow up PR

Reply:
Ok, makes sense. 6a5e41e6